controller_manager_tests/CMakeLists.txt

```diff
@@ -2,56 +2,77 @@ cmake_minimum_required(VERSION 2.8.3)
 project(controller_manager_tests)
 
 # Load catkin and all dependencies required for this package
-find_package(catkin REQUIRED COMPONENTS controller_manager controller_interface)
+find_package(catkin REQUIRED COMPONENTS
+  controller_interface
+  controller_manager
+  hardware_interface
+  pluginlib
+  roscpp
+)
+
 catkin_python_setup()
 
-include_directories(include ${Boost_INCLUDE_DIR} ${catkin_INCLUDE_DIRS})
+# Declare a catkin package
+catkin_package()
 
-catkin_package(
-  CATKIN_DEPENDS controller_manager controller_interface
-  INCLUDE_DIRS include
-  LIBRARIES ${PROJECT_NAME}
-  )
-
-#common commands for building c++ executables and libraries
-add_library(${PROJECT_NAME}
-  src/effort_test_controller.cpp
-  src/extensible_controllers.cpp
-  src/my_dummy_controller.cpp
-  src/my_robot_hw.cpp
-  src/pos_eff_controller.cpp
-  src/pos_eff_opt_controller.cpp
-  src/vel_eff_controller.cpp
-)
-target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES})
 
-add_executable(dummy_app src/dummy_app.cpp)
-target_link_libraries(dummy_app ${PROJECT_NAME} ${catkin_LIBRARIES})
+#############
+## Testing ##
+#############
 
 if(CATKIN_ENABLE_TESTING)
   find_package(rostest REQUIRED)
-  add_executable(cm_test test/cm_test.cpp)
-  add_dependencies(tests cm_test)
-  target_link_libraries(cm_test ${GTEST_LIBRARIES} ${catkin_LIBRARIES})
-  add_rostest(test/cm_test.test)
+
+  # Specify header include paths
+  include_directories(include ${catkin_INCLUDE_DIRS})
+
+  add_library(${PROJECT_NAME}
+    src/effort_test_controller.cpp
+    src/extensible_controllers.cpp
+    src/my_dummy_controller.cpp
+    src/my_robot_hw.cpp
+    src/pos_eff_controller.cpp
+    src/pos_eff_opt_controller.cpp
+    src/vel_eff_controller.cpp
+  )
+  target_link_libraries(${PROJECT_NAME} ${catkin_LIBRARIES})
+
+  add_executable(dummy_app EXCLUDE_FROM_ALL src/dummy_app.cpp)
```

Comment:
Where did `EXCLUDE_FROM_ALL` come from?

Reply:
When looking at the targets in all the test packages in https://github.com/ros/ros_comm/tree/melodic-devel/test as an example, I saw they marked targets intended for tests as `EXCLUDE_FROM_ALL`. Similarly, targets generated by `catkin_add_gtest()` and `catkin_add_gmock()` enable the property.

I admit, I don't fully understand why this is required, since the test targets are already wrapped in the `if(CATKIN_ENABLE_TESTING)`. I have the impression it was to cover an edge-case with `catkin_make` (as opposed to `catkin_make_isolated` or `catkin_tools`), something to do with the single CMake invocation over the entire workspace, but instead of properly researching and understanding it, I just used those packages as examples and mimicked.

Reply:
I think you have to add the targets to `tests` manually:
https://github.com/ros/ros_comm/blob/8d9af410949423342150b79e763348f963e19734/test/test_roscpp/test/src/CMakeLists.txt#L226

Reply:
Or use `add_rostest_gtest`, which would [automate](https://github.com/ros/ros_comm/blob/melodic-devel/tools/rostest/cmake/rostest-extras.cmake.em#L67) most of the steps.

Reply:
> I think you have to add the targets to `tests` manually

Ah, I think you're right, that makes sense. Since those targets are only depended on by targets in `add_rostest`, not `add_rostest_gtest`, they aren't directly or transitively depended on by the `tests` target, and since I added the `EXCLUDE_FROM_ALL` property, they aren't part of the `all` target either. Sounds like this may be the cause of [the colcon failure you mentioned](https://github.com/ros-controls/ros_control/issues/410#issuecomment-574397248).

I will re-add the `add_dependencies(tests target_name)` for executable targets used in tests. What are your and @bmagyar's thoughts on also adding library targets as dependencies? Back to the `ros_comm` test package examples, they seem to only mark executables as deps, and rely on libraries being pulled in transitively.

Reply:
> rely on libraries being pulled in transitively.

This is sufficient. Add direct dependencies only.